presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeyValuePairs.java

```diff
@@ -103,4 +118,63 @@ public void add(Block key, Block value, int keyPosition, int valuePosition)
             }
         }
     }
+
+    private boolean keyExists(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        return keyPositionByHash[getHashPositionOfKey(key, position)] != EMPTY_SLOT;
+    }
+
+    private void addKey(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        keyType.appendTo(key, position, keyBlockBuilder);
+        int hashPosition = getHashPositionOfKey(key, position);
+        if (keyPositionByHash[hashPosition] == EMPTY_SLOT) {
+            keyPositionByHash[hashPosition] = keyBlockBuilder.getPositionCount() - 1;
+            if (keyBlockBuilder.getPositionCount() >= maxFill) {
+                rehash(maxFill * 2);
+            }
+        }
+    }
+
+    private int getHashPositionOfKey(Block key, int position)
+    {
+        int hashPosition = getMaskedHash(hashPosition(keyType, key, position));
+        while (true) {
+            if (keyPositionByHash[hashPosition] == EMPTY_SLOT) {
+                return hashPosition;
+            }
+            else if (positionEqualsPosition(keyType, keyBlockBuilder, keyPositionByHash[hashPosition], key, position)) {
+                return hashPosition;
+            }
+            hashPosition = getMaskedHash(hashPosition + 1);
+        }
+    }
+
+    private void rehash(int size)
+    {
+        int newHashSize = arraySize(size + 1, FILL_RATIO);
```

Comment:
Don't recalculate the size here with arraySize. You should be able to just double it. Please see how `BigintGroupByHash` does it.